### PR TITLE
Move logic from executable to lib

### LINKED ISSUE #6: bin code is skipped by coveralls. refactor into lib
The logic in [bin/xlint](https://github.com/mobile-qa/xlint/blob/master/bin/xlint) isn't covered by coveralls because it's not in lib.

See a similar commit from gergich:

> * move guts of bin/gergich so we can get coverage metrics for it
>
> https://github.com/instructure/gergich/commit/07acf1dc48516869de75d7fcf704cd5ed430ea10

### PR DESCRIPTION
**Pull request by rolyatwilson**
resolves #6

### CODE REVIEW

File: lib/xlint.rb [hunk 1]
```diff
@@ -6,6 +6,36 @@
...
+      diff = Xlint.parse_git(File.read(ARGV[0]))
+      diff.files.each do |file|
+        patch = diff.find_patch_by_file(file)
+        changes = Xlint.patch_body_changes(patch.body, file)
+        @comments.concat(Xlint.find_offenses(changes))
+      end
+    end
+
+    def save_draft
+      unless @comments.empty?
```
Comment (bootstraponline) [inline lib/xlint.rb]: `return if @comments.empty?`

File: lib/xlint.rb [hunk 2]
```diff
@@ -6,6 +6,36 @@
...
+        raise 'gergich comment command failed!' unless system("gergich comment #{Shellwords.escape(@comments.to_json)}")
+      end
+    end
+
+    def publish_draft
+      (raise 'gergich publish command failed!' unless system('gergich publish')) unless @comments.empty?
```
Comment (bootstraponline) [inline lib/xlint.rb]: `return if @comments.empty?`

File: lib/xlint.rb [hunk 3]
```diff
@@ -6,6 +6,36 @@
...
+      diff.files.each do |file|
+        patch = diff.find_patch_by_file(file)
+        changes = Xlint.patch_body_changes(patch.body, file)
+        @comments.concat(Xlint.find_offenses(changes))
+      end
+    end
+
+    def save_draft
+      return if @comments.empty?
+      raise 'gergich comment command failed!' unless system("gergich comment #{Shellwords.escape(@comments.to_json)}")
```
Comment (bootstraponline) [inline lib/xlint.rb]:
```ruby
success = system("gergich comment #{Shellwords.escape(@comments.to_json)}")
raise 'gergich comment command failed!' unless success
```

File: spec/xlint_spec.rb [hunk 4]
```diff
@@ -1,11 +1,120 @@
...
+      end
+
+      it 'raises error when gerrit_base_url not set' do
+        Xlint.instance_variable_set(:@comments, good_comments)
+        expect { Xlint.publish_draft }.to raise_error(RuntimeError)
+      end
+
+      it 'raises error when gerrit_base_url not set' do
+        ENV['GERRIT_BASE_URL'] = 'someBase'
+        Xlint.instance_variable_set(:@comments, good_comments)
```
Comment (bootstraponline) [inline spec/xlint_spec.rb]: please don't use `instance_variable_set`, we should have a proper accessor.

File: lib/xlint.rb (hunk 3 above)
Comment (rolyatwilson) [inline lib/xlint.rb]: great suggestion! much cleaner and easier to read

File: lib/xlint.rb [hunk 5]
```diff
@@ -6,6 +6,40 @@
 
 class Xlint
   class << self
+    attr_accessor :comments
+
+    def check_args
+      # reads git diff from file, file name in ARGV[0]
+      raise ArgumentError, 'usage: xlint path/to/some.diff' unless ARGV.size == 1
```
Comment (bootstraponline) [inline lib/xlint.rb]: we should check that the file exists.

```
diff_file = ARGV.first
raise "File doesn't exist: #{diff_file}" unless File.exist?(diff_file)
```

File: bin/xlint [hunk 6]
```diff
@@ -2,20 +2,8 @@
...
-
-unless comments.empty?
-  raise 'gergich comment command failed!' unless system("gergich comment #{Shellwords.escape(comments.to_json)}")
-  raise 'gergich publish command failed!' unless system('gergich publish')
-end
+Xlint.check_args
+Xlint.check_env
+Xlint.build_draft
+Xlint.save_draft
+Xlint.publish_draft
```
Comment (bootstraponline) [inline bin/xlint]: `Xlint.publish` should call all these methods.

```
# in xlint.rb
def publish
  check_args
  check_env
  build_draft
  save_draft
  publish_draft
end
```

File: lib/xlint.rb [hunk 7]
```diff
@@ -6,6 +6,40 @@
...
+      diff = Xlint.parse_git(File.read(ARGV[0]))
+      diff.files.each do |file|
+        patch = diff.find_patch_by_file(file)
+        changes = Xlint.patch_body_changes(patch.body, file)
+        @comments.concat(Xlint.find_offenses(changes))
+      end
+    end
+
+    def save_draft
+      return if @comments.empty?
```
Comment (bootstraponline) [inline lib/xlint.rb]: comments is now an attribute so we don't need `@` for reading the value.

File: spec/xlint_spec.rb [hunk 8]
```diff
@@ -1,11 +1,120 @@
...
+  let(:message) { 'Deployment target changes should be approved by the team lead.' }
+  let(:severity) { 'error' }
+
+  before(:each) do
+    ARGV.clear
+    ENV.clear
+  end
+
+  after(:each) do
+    ARGV.clear
```
Comment (bootstraponline) [inline spec/xlint_spec.rb]: we don't need to clear on both before & after. before is sufficient.

File: spec/xlint_spec.rb [hunk 9]
```diff
@@ -1,11 +1,120 @@
...
+
+  context 'gergich commands' do
+    let(:bad_comments) { { someKey: 'someValue' } }
+    let(:good_comments) { { path: 'somePath', position: 1234, message: 'someMessage', severity: 'error' } }
+    let(:comment_error) { 'gergich comment command failed!' }
+    let(:publish_error) { 'gergich publish command failed!' }
+
+    describe 'save_draft' do
+      it 'does not raise error if comments are empty' do
+        Xlint.comments = []
```
Comment (bootstraponline) [inline spec/xlint_spec.rb]: shouldn't we clear the comments in `before(:each)`?

File: lib/xlint.rb [hunk 10]
```diff
@@ -3,9 +3,60 @@
...
+        patch = diff.find_patch_by_file(file)
+        changes = Xlint.patch_body_changes(patch.body, file)
+        @comments.concat(Xlint.find_offenses(changes))
+      end
+    end
+
+    def save_draft
+      return if comments.empty?
+      draft = Gergich::Draft.new
+      @comments.each do |comment|
```
Comment (bootstraponline) [inline lib/xlint.rb]: `comments.each`

**Comment by bootstraponline** [comment]
looks good to me 👍

**Pull request merged**